clients/spark/core/src/test/scala/io/treeverse/utils/AsInterfaceTest.scala

```diff
@@ -0,0 +1,71 @@
+package io.treeverse.utils
+
+import org.scalatest.funsuite.AnyFunSuite
+
+import org.objectweb.asm.{ClassReader, ClassWriter}
+import org.objectweb.asm.util.CheckClassAdapter
+
+import java.io.ByteArrayOutputStream
+import com.google.common.io.ByteStreams
+
+class AsInterfaceTest extends AnyFunSuite {
+  test("[internal] Boo does not implement Foo") {
+    assertThrows[ClassCastException] {
+      (new Boo(9)).asInstanceOf[Foo]
+    }
+  }
+
+  test("AsInterface: Boo as Foo") {
+    val booStream = this.getClass.getClassLoader.getResourceAsStream("io/treeverse/utils/Boo.class")
+    val booBytes = new ByteArrayOutputStream
+    ByteStreams.copy(booStream, booBytes)
+
+    val cr = new ClassReader(booBytes.toByteArray)
+    val cw = new ClassWriter(cr, 0)
+    val cc = new CheckClassAdapter(cw);
+    val ai = new AsInterface(cc, Map(("Foo", classOf[Foo])))
+    cr.accept(ai, 0)
+  }
+
+  test("Boo as Foo") {
+    val aicl = new AsInterfaceClassLoader(
+      getClass.getClassLoader,
+      Map(("io/treeverse/utils/Boo", classOf[Foo])))
+    val f = aicl.newInstance[Foo]("io.treeverse.utils.Boo", 9)
+    assert(f.foo(4) === 36)
+  }
+
+  test("Boo as Foo can use method taking and returning a Boo") {
+    val aicl = new AsInterfaceClassLoader(
+      getClass.getClassLoader,
+      Map(("io/treeverse/utils/Boo", classOf[Foo])))
+    val f: Foo = aicl.newInstance[Foo]("io.treeverse.utils.Boo", 9)
+    val f2: Foo = aicl.newInstance[Foo]("io.treeverse.utils.Boo", 3)
+    assert(f.addFoo(f2).foo(4) === (9 + 3) * 4)
+  }
+
+  test("Goo as Foo fails to call missing method at call time :-(") {
```

Comment:
Why at call time? Can't we recognise the missing method during mapping or instance creation?

Reply:
Unfortunately not: it may be defined in a superclass, and there is no way to verify it at translation time.

In order to do it at construction time we would need to store all the required methods somewhere and test (that would work, but might be surprising).  In order to do it at translation time (which would give the best developer experience) we would need to load the entire class into *another* ClassLoader, and then go over its public methods.

We decided F2F to delay doing this.  If you now think it's too bad an experience (and I can certainly see why), let me know & I shall add something.